engine/lib/pagehandler.php

```diff
@@ -291,17 +293,74 @@ function elgg_error_page_handler($hook, $type, $result, $params) {
 }
 
 /**
+ * Prepares a successful response to be returned by a page or an action handler
+ *
+ * @param mixed  $content     Response content
+ *                            In page handlers, response content should contain an HTML string
+ *                            In action handlers, response content can contain either a JSON string or an array of data
+ * @param string $message     System message visible to the client
+ *                            Can be used by handlers to display a system message
+ * @param string $forward_url Forward URL
+ *                            Can be used by handlers to redirect the client on non-ajax requests
+ * @param int    $status_code HTTP status code
+ *                            Status code of the HTTP response (defaults to 200)
+ * @return \Elgg\Http\OkResponse
+ */
+function elgg_ok_response($content = '', $message = '', $forward_url = null, $status_code = ELGG_HTTP_OK) {
+	if ($message) {
+		system_message($message);
+	}
+	return new \Elgg\Http\OkResponse($content, $status_code, $forward_url);
+	
+}
+
+/**
+ * Prepare an error response to be returned by a page or an action handler
+ *
+ * @param string $error       Error message
+ *                            Can be used by handlers to display an error message
+ *                            For certain requests this error message will also be used as the response body
+ * @param string $forward_url URL to redirect the client to
+ *                            Can be used by handlers to redirect the client on non-ajax requests
+ * @param int    $status_code HTTP status code
+ *                            Status code of the HTTP response
+ *                            For BC reasons and due to the logic in the client-side AJAX API,
+ *                            this defaults to 200. Note that the Router and AJAX API will
+ *                            treat these responses as error in spite of the HTTP code assigned
+ * @return \Elgg\Http\ErrorResponse
+ */
+function elgg_error_response($error = '', $forward_url = REFERRER, $status_code = ELGG_HTTP_OK) {
```

Comment:
We decided that not every error is an HTTP error, e.g. you had objections about using HTTP codes for form validation errors. That was also supposed to help with BC if we start using this in actions (thought I think the code has changed a lot since, and we ensure BC in other ways).